File: src/main/kotlin/uk/ac/ic/matterialize/matting/Matter.kt
package uk.ac.ic.matterialize.matting

import org.opencv.core.Core
import org.opencv.core.Mat
import org.opencv.core.Scalar

interface Matter {

    fun backgroundMask(videoFrame: Mat): Mat

    fun greenscreen(videoFrame: Mat): Mat {
        val greenscreen = Mat(videoFrame.size(), videoFrame.type(), Scalar(0.0, 255.0, 0.0))
        return changeBackground(videoFrame, greenscreen)
    }

    fun changeBackground(videoFrame: Mat, newBackground: Mat): Mat {
        val bgMask = backgroundMask(videoFrame)
        val fgMask = Mat()
        Core.bitwise_not(bgMask, fgMask)
        val bg = Mat()
        val fg = Mat()
        videoFrame.copyTo(fg, bgMask)
        newBackground.copyTo(bg, fgMask)

        val result = Mat()
        Core.add(fg, bg, result)

        return result
    }

    fun removeBackground(videoFrame: Mat): Mat {
        val result = Mat()
        val fgMask = backgroundMask(videoFrame)
        videoFrame.copyTo(result, fgMask)

        return result
    }
}

File: src/files/matting/matter.cpp
#include "matter.hpp"

#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>

cv::Mat IMatter::change_background(
    const cv::Mat &video_frame, const cv::Mat &new_background) {
  const cv::Mat &bg_mask{background_mask(video_frame)};
  cv::Mat fg_mask(bg_mask.size(), bg_mask.type());

  cv::bitwise_not(bg_mask, fg_mask);

  cv::Mat bg{};
  cv::Mat fg{};

  new_background.copyTo(bg, fg_mask);
  video_frame.copyTo(fg, bg_mask);

  cv::Mat result(fg.size(), fg.type());

  cv::add(fg, bg, result);

  return result;
}

cv::Mat IMatter::remove_background(const cv::Mat &video_frame) {
  const cv::Mat &fg_mask{background_mask(video_frame)};
  cv::Mat result{};

  video_frame.copyTo(result, fg_mask);

  return result;
}

cv::Mat IMatter::green_screen(const cv::Mat &video_frame) {
  const cv::Scalar scalar(0.0, 255.0, 0.0);
  const cv::Mat green_screen(video_frame.size(), video_frame.type(), scalar);

  return change_background(video_frame, green_screen);
}
